src-electron/generator/helper-endpointconfig.js

```diff
@@ -112,6 +113,23 @@ function endpoint_fixed_parent_id_array(options) {
   return '{ ' + parentIds.join(', ') + ' }'
 }
 
+async function endpoint_composition() {
+  let dt = []
+  let db = this.global.db
+  for (const et of this.endpointTypes) {
+    const result = await queryDeviceType.selectDeviceTypeById(
+      db,
+      et.deviceTypeRef
+    )
+    if (result.composition == null) {
+      dt.push('0')
+    } else {
+      dt.push(result.composition)
+    }
+  }
+  return '{' + dt.join(', ') + ' }'
```

Comment:
We should not be returning strings. This limits how the helper can be used in various ways. Let the template handle the string formation. See other block helpers and how they return a map(db-mapping.js or a custom map). If this is just about retrieving a composition of each endpoint then you can add composition to the existing endpoint helpers. It will be more useful that way.

Reply:
ill think about this and get back to you thanks

Reply:
fixed, thanks